pkg/dyninst/integration_test.go

```diff
@@ -0,0 +1,198 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+package dyninst_test
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/ringbuf"
+	"github.com/stretchr/testify/require"
+
+	"github.com/DataDog/datadog-agent/pkg/dyninst/compiler"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/config"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/ir"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/irgen"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/irprinter"
+	object "github.com/DataDog/datadog-agent/pkg/dyninst/obgect"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/output"
+	"github.com/DataDog/datadog-agent/pkg/dyninst/testprogs"
+	"github.com/DataDog/datadog-agent/pkg/util/safeelf"
+)
+
+func TestDyninst(t *testing.T) {
+	for _, cfg := range testprogs.CommonConfigs {
+		if cfg.GOARCH != runtime.GOARCH {
+			continue
+		}
+		t.Run(cfg.String(), func(t *testing.T) {
+			bin, err := testprogs.GetBinary("events_simple", cfg)
+			require.NoError(t, err)
+			testDyninst(t, bin)
+		})
+	}
+}
+
+func testDyninst(t *testing.T, sampleServicePath string) {
+	t.Logf("loading binary")
+	tempDir, err := os.MkdirTemp(os.TempDir(), "dyninst-integration-test-")
+	require.NoError(t, err)
+	defer func() {
+		if t.Failed() {
+			t.Logf("leaving temp dir %s for inspection", tempDir)
+		} else {
+			require.NoError(t, os.RemoveAll(tempDir))
+		}
+	}()
+
+	// Load the binary and generate the IR.
+	binary, err := safeelf.Open(sampleServicePath)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, binary.Close()) }()
+
+	probes := []config.Probe{
+		&config.LogProbe{
+			ID: "intArg",
+			Where: &config.Where{
+				MethodName: "main.intArg",
+			},
+		},
+	}
+
+	obj, err := object.NewElfObject(binary)
+	require.NoError(t, err)
+	require.NotNil(t, obj)
```

Comment:
nit: I find a lot of this `NotNil` checking pedantic and noisy -- other stuff is going to fall apart if things are nil. Technically not wrong, but I wouldn't include most of them.